web/src/core/mantle/evaluator/simple/index.ts

```diff
@@ -76,30 +84,41 @@ export function clearAllExpressionCaches(
   feature: Feature | undefined,
 ) {
   const appearances: Partial<LayerAppearanceTypes> = pick(layer, appearanceKeys);
-
   Object.entries(appearances).forEach(([, v]) => {
-    Object.entries(v).forEach(([, expressionContainer]) => {
-      if (hasExpression(expressionContainer)) {
-        const styleExpression = expressionContainer.expression;
-        if (typeof styleExpression === "object" && styleExpression.conditions) {
-          styleExpression.conditions.forEach(([expression1, expression2]) => {
-            clearExpressionCaches(expression1, feature, layer?.defines);
-            clearExpressionCaches(expression2, feature, layer?.defines);
-          });
-        } else if (typeof styleExpression === "boolean" || typeof styleExpression === "number") {
-          clearExpressionCaches(String(styleExpression), feature, layer?.defines);
-        } else if (typeof styleExpression === "string") {
-          clearExpressionCaches(styleExpression, feature, layer?.defines);
-        }
+    recursiveClear(v, layer, feature);
+  });
+}
+
+function recursiveClear(obj: any, layer: LayerSimple | undefined, feature: Feature | undefined) {
+  Object.entries(obj).forEach(([, v]) => {
+    // if v is an object itself and not a null, recurse deeper
+    if (hasNonExpressionObject(v)) {
+      recursiveClear(v, layer, feature);
+    } else if (hasExpression(v)) {
+      // if v is not an object, apply the clearExpressionCaches function
+      const styleExpression = v.expression;
+      if (typeof styleExpression === "object" && styleExpression.conditions) {
+        styleExpression.conditions.forEach(([expression1, expression2]) => {
+          clearExpressionCaches(expression1, feature, layer?.defines);
+          clearExpressionCaches(expression2, feature, layer?.defines);
+        });
+      } else if (typeof styleExpression === "boolean" || typeof styleExpression === "number") {
+        clearExpressionCaches(String(styleExpression), feature, layer?.defines);
+      } else if (typeof styleExpression === "string") {
+        clearExpressionCaches(styleExpression, feature, layer?.defines);
       }
-    });
+    }
   });
 }
 
 function hasExpression(e: any): e is ExpressionContainer {
   return typeof e === "object" && e && "expression" in e;
 }
 
+function hasNonExpressionObject(v: any): boolean {
+  return typeof v === "object" && v !== null && !("expression" in v);
```

Comment:
```suggestion
  return typeof v === "object" && v && !("expression" in v);
```
Sorry, it's really a small point, but it will confuse to me. So I think it should use the same code with `hasExpression` function.

Reply:
Oh no, I understand.